src/python/pants/jvm/jdk_rules.py

```diff
@@ -4,38 +4,49 @@
 from __future__ import annotations
 
 import dataclasses
+import logging
 import os
 import re
 import shlex
 import textwrap
 from dataclasses import dataclass
 from typing import ClassVar, Iterable, Mapping
 
+from pants.base.deprecated import warn_or_error
 from pants.engine.fs import CreateDigest, Digest, FileContent, FileDigest, MergeDigests
 from pants.engine.internals.selectors import Get
 from pants.engine.platform import Platform
 from pants.engine.process import BashBinary, FallibleProcessResult, Process, ProcessCacheScope
 from pants.engine.rules import collect_rules, rule
+from pants.engine.target import CoarsenedTarget
 from pants.jvm.compile import ClasspathEntry
 from pants.jvm.resolve.common import Coordinate, Coordinates
 from pants.jvm.resolve.coursier_fetch import CoursierLockfileEntry
 from pants.jvm.resolve.coursier_setup import Coursier
 from pants.jvm.subsystems import JvmSubsystem
+from pants.jvm.target_types import JvmCompatibleJdkVersionField
 from pants.util.frozendict import FrozenDict
 from pants.util.logging import LogLevel
 from pants.util.meta import frozen_after_init
 
+logger = logging.getLogger(__name__)
+
 
 @dataclass(frozen=True)
 class Nailgun:
     classpath_entry: ClasspathEntry
 
 
 @dataclass(frozen=True)
-class JdkRequest:
+class JdkRequest_:
     """Request for a JDK with a specific major version."""
 
-    version: str
+    version: str | None
+
+
+@dataclass(frozen=True)
+class JdkRequest:
+    jdk_field: JvmCompatibleJdkVersionField
```

Comment:
Since this is already `| None`, consider making it a tri-state union of `version: str | JvmCompatibleJdkVersionField | None`?